source/agora/consensus/protocol/Nominator.d

```diff
@@ -646,27 +646,26 @@ extern(D):
                     envelope.statement.pledges.confirm_.ballot.value, ex);
                 return;
             }
+            const validators = last_block.header.validators.count();
+            auto signed = last_block.header.validators.setCount;
+            if (last_block.header.height > 1 // Genesis block is not signed
+                && signed <= validators / 2) // Not more than half have signed previous block
+            {
+                log.info("Waiting for more than half to sign the last block, signed={}",
+                    last_block.header.validators.setCount);
+                this.queued_envelopes.insertBack(envelope.serializeFull.deserializeFull!SCPEnvelope); // Put back in queue
```

Comment:
this is actually dangerous now, it could put us in a busy loop in `envelopeProcessTask` without ever context switching. Maybe we should yield every time we requeue an envelope?

Reply:
That is why I used the longer interval to rearm the envelope_timer in the next line. Although ideally we should only reprocess this confirm envelope when we have the signatures. How about a second queue for commit messages pending signatures or tx?

Reply:
> That is why I used the longer interval to rearm the envelope_timer in the next line. 

But the timer handler is already running, so the `while` loop in `envelopeProcessTask` may never terminate.

> How about a second queue for commit messages pending signatures or tx?

Yeah that could work but I would've avoided doing that just to KISS and not to create any weird edge cases.

Reply:
I have taken a different approach by using the `ValidationLevel.kMaybeValidValue` status if this node does not have enough signatures for the last block when `validateValue` is called from SCP.